File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """
        GANG GANG
        """

        max_profit = 0
        min_price_seen = float("inf")

        for i in range(len(prices)):
            if prices[i] <= min_price_seen:
                min_price_seen = prices[i]
            
            max_profit = max(max_profit, prices[i] - min_price_seen)


        return max_profit

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-1.cpp
class Solution {
public:
    int BUY = 0;
    int SELL = 1;

    vector<vector<int>>cache;
    int dfs(vector<int>&prices,int curr,int prev_decision,int n){
        if(curr >= n){
            return 0;
        }

        if(cache[prev_decision][curr] != -1){
            return cache[prev_decision][curr];
        }

        int result = 0;
        int skip = dfs(prices,curr+1,prev_decision,n);
        if(prev_decision == BUY){
            int buy = max(skip,dfs(prices,curr+1,SELL,n) - prices[curr]);
            result = max(result,buy);
        }
        else{
            int sell = max(skip,dfs(prices,curr+2,BUY,n) + prices[curr]);
            result = max(result,sell);
        }

        return cache[prev_decision][curr] = result;
    }
    int maxProfit(vector<int>& prices) {
        /*
            BUY(0)
            profit = 1 
        */
        int n = prices.size();
        cache.clear();
        cache.resize(2,vector<int>(n+1,-1));
        return dfs(prices,0,BUY,n);
    }
};
